doc/guides/zephyr_cmake_package.rst

```diff
@@ -0,0 +1,418 @@
+   ├── prj.conf
+   └── src
+       └── main.c
+
+In this example ``app`` is a Zephyr freestanding application.
+
+
+Zephyr Base Environment Setting
+*******************************
+
+The Zephyr CMake package search functionality allows for explicitly specifying
+a Zephyr base using an environment variable.
+
+To do this, use the following ``find_package()`` syntax:
+
+.. code-block:: cmake
+
+   find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
+
+This syntax instructs CMake to first search for Zephyr at the location pointed
+to by the :envvar:`ZEPHYR_BASE` environment variable then use the normal search
+paths.
+
+.. _zephyr_cmake_search_order:
+
+Zephyr CMake Package Search Order
+*********************************
+
+When ``ZEPHYR_BASE`` is not used for searching, the Zephyr installation matching
+the following criteria will be used:
+
+* A Zephyr repository application will use the Zephyr in which it is located.
+  For example:
+
+  .. code-block:: none
+
+        <projects>/zephyr-workspace/zephyr
+        └── samples
+            └── hello_world
+
+  in this example, ``hello_world`` will use ``<projects>/zephyr-workspace/zephyr``.
+
+
+* Zephyr workspace application will use the Zephyr that share the same workspace.
+  For example:
+
+  .. code-block:: none
+
+     <projects>/zephyr-workspace
+     ├── zephyr
+     ├── ...
+     └── my_applications
+          └── my_first_app
+
+  in this example, ``my_first_app`` will use ``<projects>/zephyr-workspace/zephyr`` as this Zephyr
+  is located in the same workspace as the Zephyr workspace application.
+
+
+* Zephyr freestanding application will use the Zephyr registered in the CMake user package registry.
+  For example:
+
+  .. code-block:: none
+
+     <projects>/zephyr-workspace-1
+     └── zephyr                       (Not exported to CMake)
+
+     <projects>/zephyr-workspace-2
+     └── zephyr                       (Exported to CMake)
+
+     <home>/app
+     ├── CMakeLists.txt
+     ├── prj.conf
+     └── src
+         └── main.c
+
+  in this example, only ``<projects>/zephyr-workspace-2/zephyr`` is exported to the CMake package
+  registry and therefore this Zephyr will be used by the Zephyr freestanding application
+  ``<home>/app``.
+
+  If user wants to test the application with ``<projects>/zephyr-workspace-1/zephyr``, this can be
+  done by setting ``ZEPHYR_BASE=<projects>/zephyr-workspace-1/zephyr`` in environment, before
+  running CMake.
+
+  .. note::
+
+     The Zephyr package selected on the first CMake invocation will be used for all subsequent
+     builds. To change the Zephyr package, as example to test the application using environment
+     ``ZEPHYR_BASE`` then it is neccessary to do a pristine build first
+     (See :ref:`application_rebuild`).
+
+Zephyr CMake Package Version (Zephyr Freestanding Application)
+**************************************************************
+
+When writing an out-of-tree application then it is possible to specify a Zephyr version number
```

Comment:
Not for repository or workspace applications?

Reply:
It can be used for repository and workspace applications also, if you read the note below:
```
. note:: It is also possible to specify a version number for both Zephyr repository applications
          and Zephyr workspace applications. This can be used to still ensure the correct Zephyr
	  version is installed in the workspace or repository, and may be useful to avoid
	  accidental builds when only part of a workspace has been updated.
```

but especially for a repository app, one may assume that the Zephyr version of the repository app, ought to be correct.
That's why this is described primarily as a Freestanding app benefit.

Reply:
Yep. I was confused by "is possible". Maybe change "it is possible" to "it can be beneficial" or something.

Reply:
Also: "out-of-tree" -> "freestanding"?

Reply:
> Yep. I was confused by "is possible". Maybe change "it is possible" to "it can be beneficial" or something.

Good point. Confuse one, and you might confuse more.
Will change a little.

> Also: "out-of-tree" -> "freestanding"?

Correct, seems I missed this one, when we decided on new terminology.
Nicely spotted.

Reply:
Rephrased a little in this section, and removed `Freestanding` from title, to make it clearer it can be used for all app, but most beneficial for freestanding apps.